modules/ding_nodelist/ding_nodelist.module

```diff
@@ -0,0 +1,2477 @@
+    $key = $delta;
+    $form['ding_nodelist_wrapper']['ding_nodelist_bottom_links']['more_links'][$key] = _ding_nodelist_links_form($key, NULL);
+  }
+
+  $form['ding_nodelist_wrapper']['ding_nodelist_bottom_links']['ding_bottom_links_more'] = array(
+    '#type' => 'submit',
+    '#value' => t('Add more link'),
+    '#submit' => array('ding_nodelist_add_bottom_link_submit'),
+    '#ajax' => array(
+      'wrapper' => 'ding_nodelist-more-link',
+      'callback' => 'ding_nodelist_js_add_links',
+      'effect' => 'fade',
+      'method' => 'replace',
+    ),
+  );
+
+  // Cache settings.
+  $form['ding_nodelist_wrapper']['cache_settings'] = array(
+    '#type' => 'fieldset',
+    '#collapsible' => TRUE,
+    '#collapsed' => TRUE,
+    '#tree' => FALSE,
+    '#title' => t('Cache settings'),
+  );
+
+  // Load cache settigns and prepare form.
+  if (user_access('nodelist cache settings')) {
+    ctools_include('plugins');
+    $simple_cache_plugin = ctools_plugin_get_info('panels', 'cache');
+    ctools_plugin_load_includes($simple_cache_plugin);
+    $pane = $form_state['pane'];
+    $settings = !empty($pane->cache['settings'])
+      ? $pane->cache['settings']
+      : array(
+        'lifetime' => variable_get('ding_nodelist_cache_default_time', NL_CACHE_LIFETIME),
+        'granularity' => 'context',
+      );
+    $cache_form = panels_simple_cache_settings_form($settings, $form_state['display'], $pane->pid);
+    // @todo: Granularity is forced to 'context' now.
+    unset($cache_form['granularity']);
+    $form['ding_nodelist_wrapper']['cache_settings'] = array_merge($form['ding_nodelist_wrapper']['cache_settings'], $cache_form);
+  }
+
+  // Restore other form elements if this is not the first form display.
+  if (isset($form_state['restore'])) {
+    $form = array_merge($form_state['original_form'], $form);
+  }
+
+  // Remove some standard controls.
+  unset($form['override_title']);
+  unset($form['override_title_text']);
+  unset($form['override_title_markup']);
+
+  return $form;
+}
+
+/**
+ * Generate a form piece for dynamic node select form.
+ *
+ * @param string $key
+ *   Unique key for this piece.
+ * @param string $type
+ *   Node type.
+ * @param array $node
+ *   Array with the node oject.
+ * @param int $weight
+ *   Piece weight.
+ *
+ * @return array
+ *   Form structure.
+ */
+function _ding_nodelist_selected_node($key, $type, $node = array(), $weight = 0) {
+  $form = array(
+    '#tree' => TRUE,
+    '#weight' => $weight,
+  );
+
+  // Set default value for 'node'
+  $node_default = (!is_null($node) && isset($node['node'])) ? $node['node'] : '';
+  $form['node'] = array(
+    '#type' => 'textfield',
+    '#title' => t('Select new node to add to list'),
+    '#title_display' => 'invisible',
+    '#autocomplete_path' => 'ding_nodelist/autocomplete/' . $type,
+    '#parents' => array('selected_nodes', $key, 'node'),
+    '#default_value' => $node_default,
+  );
+
+  $form['nid'] = array(
```

Comment:
There are three form elements which seem to deal with the same aspects: `node`, `nid` and `type`. Please add comments regarding their purpose and relationship.